halo2/src/circuit_builder.rs

```diff
@@ -310,6 +321,41 @@ impl<'a, T: FieldElement, F: PrimeField<Repr = [u8; 32]>> Circuit<F> for PowdrCi
         // |  None            |    None      |   None           |  | <-- Halo2 will put blinding factors in the last few rows
         // |  None            |    None      |   None           | /      of the witness columns.
 
+        // If we're in a later phase, augment the original phase-0 witness by calling the witgen_callback.
+        // If we're in phase 0, we already have the full witness and don't do anything.
+        let mut new_witness = Vec::new();
+        if let Some(witness) = self.witness {
+            let mut phase = 1;
+            let challenges = config
+                .challenges
+                .iter()
+                .filter_map(|(&challenge_id, challenge)| {
+                    let mut challenge_value = None;
+                    layouter.get_challenge(*challenge).map(|x| {
+                        // The current phase is the maximum of all available challenges + 1
+                        phase = max(phase, challenge.phase() + 1);
+                        // Set the challenge value. We don't return it here, because we'd get
+                        // a Value<T> and Halo2 doesn't let us convert it to an Option<T> easily...
+                        challenge_value = Some(T::from_bytes_le(&x.to_repr()))
+                    });
+                    challenge_value.map(|v| (challenge_id, v))
```

Comment:
Can't you pull this inside the map and get rid of the `let mut challenge_value`?

Reply:
I think not, see the comment above. `layouter.get_challenge(*challenge)` returns a `Value<F>`, which is a Halo2 type which has a `map` (as `Option<F>` would), but there is no way to convert it to an option... (It does have `into_option()`, but only with `pub(crate)` visibility.)

Reply:
oh my...

But maybe you could at least add the `challenge_id` in line 339 so line 341 can be removed.

Reply:
Hmm, IDK, the best I could come up with is this:

```rust
                .filter_map(|(&challenge_id, challenge)| {
                    let mut challenge_key_value_pair = None;
                    layouter.get_challenge(*challenge).map(|x| {
                        // The current phase is the maximum of all available challenges + 1
                        phase = max(phase, challenge.phase() + 1);
                        // Set the challenge value. We don't return it here, because we'd get
                        // a Value<T> and Halo2 doesn't let us convert it to an Option<T> easily...
                        challenge_key_value_pair =
                            Some((challenge_id, T::from_bytes_le(&x.to_repr())))
                    });
                    challenge_key_value_pair
                })
```

I can do that if you think it's better, but I'm not hyped about it either...